Tools/setup/arch.sh

```diff
@@ -0,0 +1,127 @@
+#! /usr/bin/env bash
+
+## Bash script to setup PX4 development environment on Arch Linux.
+## Tested on Manjaro 18.0.1.
+##
+## Installs:
+## - Common dependencies and tools for nuttx, jMAVSim
+## - NuttX toolchain (omit with arg: --no-nuttx)
+## - jMAVSim simulator (omit with arg: --no-sim-tools)
+##
+## Not Installs:
+## - Gazebo simulation
+## - FastRTPS and FastCDR
+
+INSTALL_NUTTX="true"
+INSTALL_SIM="true"
+
+# Parse arguments
+for arg in "$@"
+do
+	if [[ $arg == "--no-nuttx" ]]; then
+		INSTALL_NUTTX="false"
+	fi
+
+	if [[ $arg == "--no-sim-tools" ]]; then
+		INSTALL_SIM="false"
+	fi
+
+done
+
+# script directory
+DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+# check requirements.txt exists (script not run in source tree)
+REQUIREMENTS_FILE="requirements.txt"
+if [[ ! -f "${DIR}/${REQUIREMENTS_FILE}" ]]; then
+	echo "FAILED: ${REQUIREMENTS_FILE} needed in same directory as ubuntu.sh (${DIR})."
+	return 1
+fi
+
+echo
+echo "Installing PX4 general dependencies"
+
+sudo pacman -Sy --noconfirm --needed \
+	astyle \
+	base-devel \
+	ccache \
+	clang \
+	cmake \
+	cppcheck \
+	doxygen \
+	gdb \
+	git \
+	gnutls \
+	nettle \
+	ninja \
+	python-pip \
+	rsync \
+	shellcheck \
+	tar \
+	unzip \
+	wget \
+	zip \
+	;
+
+# Python dependencies
+echo "Installing PX4 Python3 dependencies"
+sudo pip install --upgrade pip setuptools wheel
+sudo pip install -r ${DIR}/requirements.txt
+
+
+# NuttX toolchain (arm-none-eabi-gcc)
+if [[ $INSTALL_NUTTX == "true" ]]; then
+	echo
+	echo "Installing NuttX dependencies"
+
+	sudo pacman -S --noconfirm --needed \
+		gperf \
+		vim \
+		;
+
+	# add user to dialout group (serial port access)
+	sudo usermod -aG uucp $USER
+
+	# Remove modem manager (interferes with PX4 serial port/USB serial usage).
+	sudo pacman -R modemmanager --noconfirm
```

Comment:
This is evil. I vote we don't do that.

Reply:
This is needed (and done) on ubuntu. So it may be evil, but consistently evil. Of course if the ubuntu and arch modem managers are different this could be pointless.

Reply:
I am also curious to know why this is evil.

Reply:
Because let's say you use modemmanager because you use LTE in your laptop and then the PX4 installer comes along and decides to just remove it.

Reply:
The right thing to do is to make pull requests against modemmanager and add all IDs of all Pixhawks so that it ignores them.

The 3DR USB ID is ignored since 2014:
https://github.com/freedesktop/ModemManager/commit/9447527c357a91bbd3660274c41fd1b802e4cf84

Reply:
OK, sounds feasible, do we have a definite list of vendor IDs for Pixhawk/Dronecode devices? FYI @mrpollo

Reply:
It looks like the only one not covered by the modemmanager blacklist is NXP's fmuk66-v3.

Reply:
Shall we not install modemmanager and drop an email to someone at NXP then?

Reply:
AFAIK we don't have a list of all Pixhawk/Dronecode devices.

I agree with @julianoes, let's keep the `modemmanager` package, let the errors come through, create issues, and do proper fixes. We should do the same on Ubuntu.

I agree with @MaEtUgR we can do the PRs for them, this is very straightforward, let's start by creating issues and labeling correctly, this is something someone can help contribute as well.

Reply:
I have the PR for NXP ready, I'm just waiting on @igalloway's answer to go ahead with it.

Reply:
Cool thanks, then I remove it from both scripts.

Reply:
What about the huge message telling you that you should remove the modemmanager during every upload?
```
[1206/1207] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...
```

Reply:
We should only display it if and only if there is a port available but it's already occupied by someone else.

Reply:
I put this to a test with a Pixhawk 4 and a CUAV V5 nano (both fmu-v5) and haing the modemmanager installed clearly breaks a lot of flashing attempts. Especially every attmpt where you start the script and then power the board every single time. Here's the output of me repeating this case with both board alternately, removing modemmmanager and all subsequent tries are successful. I don't consider to leave it if the chance of successful flashing is that low.
```
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 003d00233138510638393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [                    ] 3.9%
ERROR: timeout waiting for data (1 bytes)
FAILED: platforms/nuttx/CMakeFiles/upload 
cd /home/maetugr/Firmware/build/px4_fmu-v5_default && /usr/bin/python /home/maetugr/Firmware/Tools/px_uploader.py --port "/dev/serial/by-id/*-3D_Robotics_PX4_FMU_v5.x*,/dev/serial/by-id/*PX4_BL*,/dev/serial/by-id/*BL_FMU*,/dev/serial/by-id/usb-The_Autopilot*,/dev/serial/by-id/usb-Bitcraze*,/dev/serial/by-id/pci-Bitcraze*,/dev/serial/by-id/usb-Gumstix*,/dev/serial/by-id/usb-Hex_ProfiCNC*,/dev/serial/by-id/usb-UVify_FMU_BL*" /home/maetugr/Firmware/build/px4_fmu-v5_default/px4_fmu-v5_default.px4
ninja: build stopped: subcommand failed.
make: *** [Makefile:202: px4_fmu-v5] Error 1
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 003d00233138510638393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [                    ] 0.0%
ERROR: timeout waiting for data (1 bytes)
FAILED: platforms/nuttx/CMakeFiles/upload 
cd /home/maetugr/Firmware/build/px4_fmu-v5_default && /usr/bin/python /home/maetugr/Firmware/Tools/px_uploader.py --port "/dev/serial/by-id/*-3D_Robotics_PX4_FMU_v5.x*,/dev/serial/by-id/*PX4_BL*,/dev/serial/by-id/*BL_FMU*,/dev/serial/by-id/usb-The_Autopilot*,/dev/serial/by-id/usb-Bitcraze*,/dev/serial/by-id/pci-Bitcraze*,/dev/serial/by-id/usb-Gumstix*,/dev/serial/by-id/usb-Hex_ProfiCNC*,/dev/serial/by-id/usb-UVify_FMU_BL*" /home/maetugr/Firmware/build/px4_fmu-v5_default/px4_fmu-v5_default.px4
ninja: build stopped: subcommand failed.
make: *** [Makefile:202: px4_fmu-v5] Error 1
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 003d00233138510638393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [                    ] 0.0%
ERROR: timeout waiting for data (1 bytes)
FAILED: platforms/nuttx/CMakeFiles/upload 
cd /home/maetugr/Firmware/build/px4_fmu-v5_default && /usr/bin/python /home/maetugr/Firmware/Tools/px_uploader.py --port "/dev/serial/by-id/*-3D_Robotics_PX4_FMU_v5.x*,/dev/serial/by-id/*PX4_BL*,/dev/serial/by-id/*BL_FMU*,/dev/serial/by-id/usb-The_Autopilot*,/dev/serial/by-id/usb-Bitcraze*,/dev/serial/by-id/pci-Bitcraze*,/dev/serial/by-id/usb-Gumstix*,/dev/serial/by-id/usb-Hex_ProfiCNC*,/dev/serial/by-id/usb-UVify_FMU_BL*" /home/maetugr/Firmware/build/px4_fmu-v5_default/px4_fmu-v5_default.px4
ninja: build stopped: subcommand failed.
make: *** [Makefile:202: px4_fmu-v5] Error 1
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 002000223138510138393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [                    ] 0.0%
ERROR: timeout waiting for data (1 bytes)
FAILED: platforms/nuttx/CMakeFiles/upload 
cd /home/maetugr/Firmware/build/px4_fmu-v5_default && /usr/bin/python /home/maetugr/Firmware/Tools/px_uploader.py --port "/dev/serial/by-id/*-3D_Robotics_PX4_FMU_v5.x*,/dev/serial/by-id/*PX4_BL*,/dev/serial/by-id/*BL_FMU*,/dev/serial/by-id/usb-The_Autopilot*,/dev/serial/by-id/usb-Bitcraze*,/dev/serial/by-id/pci-Bitcraze*,/dev/serial/by-id/usb-Gumstix*,/dev/serial/by-id/usb-Hex_ProfiCNC*,/dev/serial/by-id/usb-UVify_FMU_BL*" /home/maetugr/Firmware/build/px4_fmu-v5_default/px4_fmu-v5_default.px4
ninja: build stopped: subcommand failed.
make: *** [Makefile:202: px4_fmu-v5] Error 1
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 003d00233138510638393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [                    ] 0.0%
ERROR: timeout waiting for data (1 bytes)
FAILED: platforms/nuttx/CMakeFiles/upload 
cd /home/maetugr/Firmware/build/px4_fmu-v5_default && /usr/bin/python /home/maetugr/Firmware/Tools/px_uploader.py --port "/dev/serial/by-id/*-3D_Robotics_PX4_FMU_v5.x*,/dev/serial/by-id/*PX4_BL*,/dev/serial/by-id/*BL_FMU*,/dev/serial/by-id/usb-The_Autopilot*,/dev/serial/by-id/usb-Bitcraze*,/dev/serial/by-id/pci-Bitcraze*,/dev/serial/by-id/usb-Gumstix*,/dev/serial/by-id/usb-Hex_ProfiCNC*,/dev/serial/by-id/usb-UVify_FMU_BL*" /home/maetugr/Firmware/build/px4_fmu-v5_default/px4_fmu-v5_default.px4
ninja: build stopped: subcommand failed.
make: *** [Makefile:202: px4_fmu-v5] Error 1
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
==========================================================================================================
WARNING: You should uninstall ModemManager as it conflicts with any non-modem serial device (like Pixhawk)
==========================================================================================================
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 002000223138510138393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [                    ] 0.0%
ERROR: timeout waiting for data (1 bytes)
FAILED: platforms/nuttx/CMakeFiles/upload 
cd /home/maetugr/Firmware/build/px4_fmu-v5_default && /usr/bin/python /home/maetugr/Firmware/Tools/px_uploader.py --port "/dev/serial/by-id/*-3D_Robotics_PX4_FMU_v5.x*,/dev/serial/by-id/*PX4_BL*,/dev/serial/by-id/*BL_FMU*,/dev/serial/by-id/usb-The_Autopilot*,/dev/serial/by-id/usb-Bitcraze*,/dev/serial/by-id/pci-Bitcraze*,/dev/serial/by-id/usb-Gumstix*,/dev/serial/by-id/usb-Hex_ProfiCNC*,/dev/serial/by-id/usb-UVify_FMU_BL*" /home/maetugr/Firmware/build/px4_fmu-v5_default/px4_fmu-v5_default.px4
ninja: build stopped: subcommand failed.
make: *** [Makefile:202: px4_fmu-v5] Error 1
[maetugr@ManjaroVM Firmware]$ sudo -R pacman modemmanager
sudo: invalid option -- 'R'
usage: sudo -h | -K | -k | -V
usage: sudo -v [-AknS] [-g group] [-h host] [-p prompt] [-u user]
usage: sudo -l [-AknS] [-g group] [-h host] [-p prompt] [-U user] [-u user] [command]
usage: sudo [-AbEHknPS] [-C num] [-g group] [-h host] [-p prompt] [-T timeout] [-u user] [VAR=value] [-i|-s] [<command>]
usage: sudo -e [-AknS] [-C num] [-g group] [-h host] [-p prompt] [-T timeout] [-u user] file ...
[maetugr@ManjaroVM Firmware]$ sudo pacman -R modemmanager
[sudo] password for maetugr: 
checking dependencies...
:: networkmanager optionally requires modemmanager: cellular network support

Packages (1) modemmanager-1.10.4-1

Total Removed Size:  9.55 MiB

:: Do you want to remove these packages? [Y/n] 
:: Processing package changes...
(1/1) removing modemmanager                                                                                                                          [###########################################################################################] 100%
:: Running post-transaction hooks...
(1/5) Reloading system bus configuration...
(2/5) Updating icon theme caches...
(3/5) Reloading system manager configuration...
(4/5) Reloading device manager configuration...
(5/5) Arming ConditionNeedsUpdate...
[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 002000223138510138393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [====================] 100.0%
Verify : [====================] 100.0%
Rebooting. Elapsed Time 28.415

[maetugr@ManjaroVM Firmware]$ make px4_fmu-v5 upload
[0/4] Performing build step for 'px4io_firmware'
ninja: no work to do.
[3/4] uploading px4
Loaded firmware for board id: 50,0 size: 1670412 bytes (80.92%), waiting for the bootloader...


Found board id: 50,0 bootloader version: 5 on /dev/serial/by-id/usb-3D_Robotics_PX4_BL_FMU_v5.x_0-if00
sn: 003d00233138510638393738
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [====================] 100.0%
Verify : [====================] 100.0%
Rebooting. Elapsed Time 28.433
```

Reply:
Fair enough, just leave it in for now then.

Reply:
When/if you remove this let me know? The ubuntu.sh removes it too, and should stay/go for same reasons.

As I understand this, it could cause a firmware update to fail. What would that look like to the user? [I'm thinking of adding this as a docs note]

Reply:
Let's leave it in for now and check on it another day.

@hamishwillee it's already covered here: https://dev.px4.io/master/en/debug/faq.html#usb-errors